pkg/auth/multiauthenticator.go

```diff
@@ -0,0 +1,67 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	"strings"
+)
+
+type IssuerWithClientID struct {
+	Issuer   string // Issuer of the token.
+	ClientID string // ClientID of the token, can be empty.
+}
+
+type multiAuthenticatorWithClientIDImpl struct {
+	authenticators map[IssuerWithClientID]Authenticator
+}
+
+func NewMultiAuthenticatorWithClientID(
+	authenticators map[IssuerWithClientID]Authenticator) MultiAuthenticatorWithClientID {
+	return &multiAuthenticatorWithClientIDImpl{
+		authenticators: authenticators,
+	}
+}
+
+// GetAuthenticator returns nil if there is no authenticator for the given issuer and client ID.
+func (m *multiAuthenticatorWithClientIDImpl) GetAuthenticator(issuer, audience string) Authenticator {
+	for issuerWithClientID, authenticator := range m.authenticators {
+		// Audience field in a claim can contain more than just the clientID.
+		// The strings.Contains check here is same as what is being used by the OIDC library
+		// which verifies the token.
+		if issuer == issuerWithClientID.Issuer && strings.Contains(audience, issuerWithClientID.ClientID) {
```

Comment:
I just realized that there is a problem with this approach where if a different clientID is a substring in the audience field as opposed to the intended clientID. I am thinking a better approach is to just iterate through all authenticators and if all fails then return "Unauthenticated" error. Will send a fix up. LMK what you think

Reply:
Yeah, that sounds fine. We could get authenticators by the issuer and iterate only through only those.